### PR TITLE
Simplify FusedLoad and CachedBufferedInput

### PR DESCRIPTION
**Pull request by oerling**
The IO pattern for Velox file readers is to register the pieces of file to be
accessed, then to access some or all of these pieces. Pieces that are in close
proximity get loaded with a single IO. The definition of close depends on the
media.

This removes the load_ member from AsyncDataCacheEntry. The cache entry no
longer references the FusedLoad that will be invoked to fill it. Instead, the
CacheInputStream  that will load a piece of file coalesced with other pieces
references the FusedLoad.

The API contract of CachePin is simplified: If findOrCreate returns an exclusive
pin, the caller is responsible for filling the entry and calling
setExclusiveToShared when done. This makes the data visible to other threads.
There is no more a concept of dataValid or ensureLoaded for shared mode pins
that have uninitialized data.

We add tests for checking a consistent after state for errors in loading data
into CachePins. AsyncDataCacheTest adds error injection in FusedLoads and
single pin loading. The code is also more representative of the intended usage
pattern.

CachedBufferedInput is adapted to use the changed AsyncDataCache API.
CacheInputStream looks for its FusedLoad instead of the FusedLoad being
recorded in a cache entry. Cache entries for FusedLoads are made right before
loading the FusedLoad. CoalesceIO.h is used systematically to detect and
coalesce coalesced loads.

ScanTracker exports TrackingData so that columns can be more accurately sorted
in different load frequency buckets.

### CODE REVIEW

**Comment by mbasmanova** [comment]
@oerling Orri, there are build failures:

```
FAILED: velox/dwio/dwrf/test/CMakeFiles/velox_dwio_dwrf_cache_input_test.dir/CacheInputTest.cpp.o 
ccache /opt/rh/gcc-toolset-9/root/bin/g++  -DBOOST_ALL_NO_LIB -DBOOST_CONTEXT_DYN_LINK -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_REGEX_DYN_LINK -DBOOST_SYSTEM_DYN_LINK -DBOOST_THREAD_DYN_LINK -DFMT_LOCALE -DFMT_SHARED -DGFLAGS_IS_A_DLL=0 -I../../. -Ivelox/dwio/dwrf/proto/../../../.. -isystem ../../velox -isystem ../../velox/external -isystem ../../velox/external/duckdb -isystem /usr/include/libdwarf -mavx2 -mfma -mavx -mf16c -mbmi2 -march=native -D USE_VELOX_COMMON_BASE -D HAS_UNCAUGHT_EXCEPTIONS -Werror -Wall -Wextra -Wno-unused        -Wno-unused-parameter        -Wno-sign-compare        -Wno-ignored-qualifiers        -Wno-implicit-fallthrough          -Wno-empty-body          -Wno-class-memaccess          -Wno-comment          -Wno-int-in-bool-context          -Wno-redundant-move          -Wno-array-bounds          -Wno-maybe-uninitialized          -Wno-unused-result          -Wno-format-overflow          -Wno-strict-aliasing          -Wno-type-limits -g   -std=gnu++17 -MD -MT velox/dwio/dwrf/test/CMakeFiles/velox_dwio_dwrf_cache_input_test.dir/CacheInputTest.cpp.o -MF velox/dwio/dwrf/test/CMakeFiles/velox_dwio_dwrf_cache_input_test.dir/CacheInputTest.cpp.o.d -o velox/dwio/dwrf/test/CMakeFiles/velox_dwio_dwrf_cache_input_test.dir/CacheInputTest.cpp.o -c ../../velox/dwio/dwrf/test/CacheInputTest.cpp
../../velox/dwio/dwrf/test/CacheInputTest.cpp:398:10: error: 'FileGroupStats' in namespace 'facebook::velox::cache' does not name a type
  398 |   cache::FileGroupStats* FOLLY_NULLABLE groupStats_ = nullptr;
      |          ^~~~~~~~~~~~~~
../../velox/dwio/dwrf/test/CacheInputTest.cpp: In member function 'void CacheTest::readLoop(const string&, int, int32_t, int32_t, int32_t, int32_t)':
../../velox/dwio/dwrf/test/CacheInputTest.cpp:324:63: error: 'groupStats_' was not declared in this scope; did you mean 'ioStats_'?
  324 |         std::make_shared<ScanTracker>("testTracker", nullptr, groupStats_);
      |                                                               ^~~~~~~~~~~
      |                                                               ioStats_
```

**Comment by oerling** [comment]
Fixed.

File: velox/common/caching/AsyncDataCache.cpp [hunk 1]
```diff
@@ -482,6 +460,10 @@ bool AsyncDataCache::allocateContiguous(
   });
 }
 
+void AsyncDataCache::incrementNew(uint64_t size) {
+  newBytes_ += size;
```
Comment (majetideepak) [inline velox/common/caching/AsyncDataCache.cpp:464]: Does this have to reset anywhere?

**Review by majetideepak — COMMENTED** [review]
@oerling I did a high-level pass and made some comments.

**Comment by mbasmanova** [comment]
> Looks like we don't have a naming convention for global variables.

@majetideepak @oerling I find it hard to keep track of these gflags sprinkled around the library. I think it would be easier to either not use these at all and pass configuration via context / constructor or consolidate all the flags in one place.

**Comment by oerling** [comment]
Of these flags, the load quantum is a proper config with a close analog in Presto. (orc max buffer size or something like that) in Hive config. This is from a config file and not from session. There are actually cases where this is set to a non-default value.

So, for this purpose we may add a field in ReaderOpts. 

 

The BufferedInputFactory and the Reader have the same lifetime, so this can also be passed to a CachedBufferedInputFactory.

 

Coalesce distance is a property of InputStream. It could conceivably have a different value for  AWS or HDFS.

**Comment by mbasmanova** [comment]
@oerling Orri, thank you for taking the time to write the detailed description for the PR. It is easy to follow and I hope it will help with reviewing the code. One question I have after reading the description is why FusedLoad is using term "fuse" and not "coalesce". It feels like using the same term consistently throughout the codebase would make it easier to read. Would it make sense to rename FusedLoad to CoalescedLoad?

File: velox/common/caching/AsyncDataCache.h [hunk 2]
```diff
@@ -59,14 +61,19 @@ struct AccessStats {
     return (now - lastUse) / (1 + numUses);
   }
 
+  void reset() {
```
Comment (mbasmanova) [inline velox/common/caching/AsyncDataCache.h:64]: Would you document this method? In particular, I wonder when would it be used.

File: velox/common/caching/AsyncDataCache.h [hunk 3]
```diff
@@ -542,6 +493,8 @@ class CacheShard {
       uint64_t size,
       folly::SemiFuture<bool>* FOLLY_NULLABLE readyFuture);
 
+  bool exists(RawFileCacheKey key);
```
Comment (mbasmanova) [inline velox/common/caching/AsyncDataCache.h:496]: This method can be const, right?

File: velox/common/caching/AsyncDataCache.h [hunk 4]
```diff
@@ -562,6 +515,8 @@ class CacheShard {
   // Adds the stats of 'this' to 'stats'.
   void updateStats(CacheStats& stats);
 
+  void appendSsdSaveable(std::vector<CachePin>& pins);
```
Comment (mbasmanova) [inline velox/common/caching/AsyncDataCache.h:518]: Is this used?

File: velox/common/caching/AsyncDataCache.h [hunk 5]
```diff
@@ -633,6 +588,9 @@ class AsyncDataCache : public memory::MappedMemory,
       uint64_t size,
       folly::SemiFuture<bool>* FOLLY_NULLABLE waitFuture = nullptr);
 
+  // Returns true if there is an entry for 'key'. Updates access time.
+  bool exists(RawFileCacheKey key);
```
Comment (mbasmanova) [inline velox/common/caching/AsyncDataCache.h:592]: Make this method const?

File: velox/common/caching/AsyncDataCache.h [hunk 6]
```diff
@@ -688,6 +646,49 @@ class AsyncDataCache : public memory::MappedMemory,
...
+  }
+
+  // Looks up a pin for each in 'keys' and skips all loading or
+  // loaded pins. Calls processPin for each exclusive
+  // pin. processPin must move its argument if it wants to use it
+  // afterwards. sizeFunc(i) returns the size of the ith item in
+  // 'keys'.
+  template <typename SizeFunc, typename ProcessPin>
+  void makePins(
+      std::vector<RawFileCacheKey>& keys,
```
Comment (mbasmanova) [inline velox/common/caching/AsyncDataCache.h:677]: Should keys be a const&?

File: velox/common/caching/AsyncDataCache.h [hunk 7]
```diff
@@ -688,6 +646,49 @@ class AsyncDataCache : public memory::MappedMemory,
...
+
+  // Updates statistics after bringing in 'bytes' worth of data that
+  // qualifies for SSD save and is not backed by SSD. Periodically
+  // triggers a background write of eligible entries to SSD.
+  void possibleSsdSave(uint64_t bytes);
+
+  // Sets a callback applied to new entries at the point where
+  // 'dataValid_' is set to true. Used for testing and can be used for
+  // e.g. checking checksums.
+  void setVerifyHook(std::function<void(const AsyncDataCacheEntry&)> hook) {
```
Comment (mbasmanova) [inline velox/common/caching/AsyncDataCache.h:662]: Like Deepak mentioned, the comment is out of date and needs updating.

File: velox/common/caching/AsyncDataCache.cpp [hunk 8]
```diff
@@ -195,19 +190,17 @@ CachePin CacheShard::initEntry(
       }
     }
   }
-  entry->touch();
+  entry->accessStats_.reset();
```
Comment (mbasmanova) [inline velox/common/caching/AsyncDataCache.cpp:193]: This code suggests that AsyncDataCacheEntry may be reused, but I don't understand how would that happen. Would you explain?

File: velox/common/caching/AsyncDataCache.cpp [hunk 9]
```diff
@@ -230,45 +223,25 @@ bool FusedLoad::loadOrFuture(folly::SemiFuture<bool>* wait) {
...
-      }
-    }
-    pins_.clear();
+  } catch (std::exception& e) {
+    error = std::current_exception();
   }
-  if (promise) {
-    promise->setValue(true);
+  if (error) {
+    // Cleanup outside of catch to avoid killing the process if more exceptions.
```
Comment (mbasmanova) [inline velox/common/caching/AsyncDataCache.cpp:240]: This is an unusual pattern. Why not wrap setEndState call in a separate try-catch?

File: velox/common/caching/AsyncDataCache.cpp [hunk 10]
```diff
@@ -499,8 +487,11 @@ std::string AsyncDataCache::toString() const {
       << " / " << maxBytes_ << " bytes\n"
       << "Miss: " << stats.numNew << " Hit " << stats.numHit << " evict "
       << stats.numEvict << "\n"
-      << " read pins " << stats.numShared << " unused prefetch "
-      << stats.numPrefetch << " Alloc Mclks " << (stats.allocClocks >> 20);
+      << " read pins " << stats.numShared << " write pins "
+      << stats.numExclusive << " unused prefetch " << stats.numPrefetch
+      << " Alloc Mclocks " << (stats.allocClocks >> 20) << " allocated pages "
```
Comment (mbasmanova) [inline velox/common/caching/AsyncDataCache.cpp:492]: typo: Mclocks

File: velox/common/caching/ScanTracker.h [hunk 11]
```diff
@@ -78,9 +79,13 @@ struct TrackingData {
   int64_t readBytes{};
   int32_t numReferences{};
   int32_t numReads{};
-  void incrementReference(uint64_t bytes) {
+  void incrementReference(uint64_t bytes, int32_t quantum = 0) {
```
Comment (mbasmanova) [inline velox/common/caching/ScanTracker.h:82]: It would be nice to document this method and explain the use of quantum parameter. I don't understand the new logic. Why would number of references depend on the load quantum?

**Review by mbasmanova — COMMENTED** [review]
@oerling Some comments and questions. Will continue reading tomorrow.

File: velox/dwio/dwrf/common/CachedBufferedInput.h [hunk 12]
```diff
@@ -92,28 +119,27 @@ class CachedBufferedInput : public BufferedInput {
...
-  // Removes the requests from 'toLoad' if they hit SSD cache. May start
-  // background load from SSD.
-  void loadFromSsd(std::vector<CacheRequest*> requests);
+  // Sorts requests and makes FusedLoads for nearby requests. If 'prefetch' is
+  // true, starts background loading.
+  void makeLoads(std::vector<CacheRequest*> requests, bool prefetch);
+
+  // Makes a FusedLoad for 'requests' to be read together, coalescing
+  // IO is appropriate. If 'prefetch' is set, schedules the FusedLoad
+  // on 'executor_'. Links the FusedLoad  to all CacheInputStreams tat it
```
Comment (mbasmanova) [inline velox/dwio/dwrf/common/CachedBufferedInput.h:140]: typos: tat -> that, concers -> concerns

File: velox/dwio/dwrf/common/CachedBufferedInput.cpp [hunk 13]
```diff
@@ -59,193 +79,321 @@ bool CachedBufferedInput::shouldPreload() {
...
   if (numPages + prefetchPages < cachePages / 2) {
+    // The planned prefetch plus other prefetches are under half the cache.
     return true;
   }
   return false;
 }
 
+namespace {
+
+std::vector<int32_t> makeReadPctBuckets() {
```
Comment (mbasmanova) [inline velox/dwio/dwrf/common/CachedBufferedInput.cpp:103]: It would be nice to document this function to explain the choice of constants.

File: velox/dwio/dwrf/common/CachedBufferedInput.cpp [hunk 14]
```diff
@@ -59,193 +79,321 @@ bool CachedBufferedInput::shouldPreload() {
...
+  return {80, 50, 20, 0};
+}
+
+const static std::vector<int32_t>& readPctBuckets() {
+  static std::vector<int32_t> buckets = makeReadPctBuckets();
```
Comment (mbasmanova) [inline velox/dwio/dwrf/common/CachedBufferedInput.cpp:108]: makeReadPctBuckets function seems unnecessary. Why can't you just initialize the vector as buckets = {80, 50, 20, 0};?

File: velox/dwio/dwrf/common/CachedBufferedInput.cpp [hunk 15]
```diff
@@ -15,14 +15,31 @@
  */
 
 #include "velox/dwio/dwrf/common/CachedBufferedInput.h"
+#include "velox/common/process/TraceContext.h"
 #include "velox/dwio/dwrf/common/CacheInputStream.h"
 
+DEFINE_int32(cache_load_quantum, 8 << 20, "Max size of single IO to cache");
+DEFINE_int32(
+    cache_prefetch_min_pct,
```
Comment (mbasmanova) [inline velox/dwio/dwrf/common/CachedBufferedInput.cpp:23]: This logic doesn't seem to be covered by PR description. Can it be moved into a separate PR?

**Review by mbasmanova — COMMENTED** [review]
@oerling Orri, I'm seeing a lot of logic related to prefetch. This is unexpected given the PR description. Is it included by mistake?

**Comment by oerling** [comment]
Could do that.

**Comment by oerling** [comment]
Updated the PR. Below are replies to comments.

 

 

Q: I'm confused by this comment suggesting that read ahead is done on query processing thread. Isn't the whole point to read ahead in background?

 

 

 

A: If you need something that is in the process of prefetching you need to wait for the prefetch to finish. If all prefetch were perfect this counter would only count IOs on the query thread.

 

 

 

+        } catch (const std::exception& e) {

+          LOG(ERROR) << "IOERR:

 

error in fused load " << e.what();

Why is it OK to ignore this error?

 

A: The fused load is a speculative operation. The individual targeted entry will be hit next. The error will be signalled when hitting the individual entry. This makes it so the error signalling code occurs only once.

 

 

 

A:

+  entry->accessStats_.reset();

This code suggests that AsyncDataCacheEntry may be reused, but I don't understand how would that happen. Would you explain?

 

A: getFreeEntryWithSize can reuse an entry. The counters should be reset. 

 

 

+    // Cleanup outside of catch to avoid killing the process if more exceptions.

This is an unusual pattern. Why not wrap setEndState call in a separate try-catch?

 

A: catches should be nothrow. Anything can in principle throw, for example getting a mutex can, even though in practice it does not. Now, if recovery is wrapped in another catch throw then whatever error that may signal is lost. So it is better to do like this. In fact this pattern should be preferred henceforth since it is more tractable.

 

 

+      << " Alloc Mclocks " << (stats.allocClocks >> 20) << " allocated pages "

typo: Mclocks

 

 

 

A: M for mega is intended.

 

 

 

:

This logic doesn't seem to be covered by PR description. Can it be moved into a separate PR?

—

 

A:  We delete the prefetchThreshold member and make a flag instead.  The point is to process the frequently loaded separate from the infrequently loaded. Coalescing infrequent with frequent is not necessarily good so better not do it. If did, there could be weird cases that defeat te whole business of selective load.

We keep a flag for this now

because  this has a break even that must be emprically determined. After that the flag can be a constant.

 

The quantum and coalesce distance are proper connector properties  from Presto. These will be set separately.

 

 

The test (CacheInputTest) produces different load frequencies and hits this  code path. The point is to test the executor with some async loads to check sync between read-ahead and actual read of the same thing. This is already in the previous state.

**Comment by mbasmanova** [comment]
@oerling Thank you for the replies.

> A: If you need something that is in the process of prefetching you need to wait for the prefetch to finish. If all prefetch were perfect this counter would only count IOs on the query thread.

That makes sense. It would be nice to clarify in the comment as well.

> A: The fused load is a speculative operation. The individual targeted entry will be hit next. The error will be signalled when hitting the individual entry. This makes it so the error signalling code occurs only once.

Understood. Please, add a comment to clarify.

> A: getFreeEntryWithSize can reuse an entry. The counters should be reset. 

Got it. In this case, it would be cleaner to call entry->reset() from getFreeEntryWithSize. The reset() would take care of resetting all state including stats. Alternatively, entry->reset() could be called first thing in initEntry(). In this case a quick comment explaining that the entry may have been recycled would help.

> A: M for mega is intended.

I wouldn't have guessed. Perhaps, change to MegaClocks for clarity.

> A: catches should be nothrow. Anything can in principle throw, for example getting a mutex can, even though in practice it does not. Now, if recovery is wrapped in another catch throw then whatever error that may signal is lost. So it is better to do like this. In fact this pattern should be preferred henceforth since it is more tractable.

I'm not sure I understand. Won't the following work?

```
catch (std::exception& e) {
  try {
    setEndState(LoadState::kCancelled);
  } catch (...) {
    // swallow the error
  }

  throw;
}
```

> A: We delete the prefetchThreshold member and make a flag instead. The point is to process the frequently loaded separate from the infrequently loaded. Coalescing infrequent with frequent is not necessarily good so better not do it. If did, there could be weird cases that defeat te whole business of selective load.

This sounds like a separate change. Can it be extracted into a separate PR?

File: velox/common/caching/ScanTracker.h [hunk 16]
```diff
@@ -78,9 +79,20 @@ struct TrackingData {
   int64_t readBytes{};
   int32_t numReferences{};
   int32_t numReads{};
-  void incrementReference(uint64_t bytes) {
+
+  // Marks that 'bytes' worth of data in the tracked object has been
+  // accessed. If 'bytes' is larger than a single access quantum, the
+  // reference counts for as many accesses tat are needed to cover
```
Comment (mbasmanova) [inline velox/common/caching/ScanTracker.h:85]: typo: tat -> that

File: velox/common/caching/ScanTracker.h [hunk 17]
```diff
@@ -78,9 +79,20 @@ struct TrackingData {
...
+  // 'bytes'. When reading a large object, we will get a read per
+  // quantum. So then if the referenced and read quanta match, we know
+  // that the object is densely read.
+  void incrementReference(uint64_t bytes, int32_t quantum = 0) {
```
Comment (mbasmanova) [inline velox/common/caching/ScanTracker.h:89]: Orri, thank you for a very nice description. Why is quantum an optional parameter with default value of 0. Shouldn't we always specify it? Would it make sense to rename it to loadQuantum for clarity? Also, quantum is constant, it would be more natural to pass it in the constructor, no?

File: velox/dwio/dwrf/common/CachedBufferedInput.h [hunk 18]
```diff
@@ -123,15 +149,19 @@ class CachedBufferedInput : public BufferedInput {
...
   std::vector<CacheRequest> requests_;
+
   // Coalesced loads spanning multiple cache entries in one IO.
-  std::vector<std::shared_ptr<cache::FusedLoad>> fusedLoads_;
+  folly::Synchronized<folly::F14FastMap<
+      const SeekableInputStream*,
+      std::shared_ptr<cache::CoalescedLoad>>>
+      fusedLoads_;
+
+  // Distinct fused loads in 'fusedLoads_'.
```
Comment (mbasmanova) [inline velox/dwio/dwrf/common/CachedBufferedInput.h:161]: fusedLoads_ -> coalescedLoads_

File: velox/dwio/dwrf/common/CachedBufferedInput.h [hunk 19]
```diff
@@ -123,15 +149,19 @@ class CachedBufferedInput : public BufferedInput {
...
-
   // Regions that are candidates for loading.
   std::vector<CacheRequest> requests_;
+
   // Coalesced loads spanning multiple cache entries in one IO.
-  std::vector<std::shared_ptr<cache::FusedLoad>> fusedLoads_;
+  folly::Synchronized<folly::F14FastMap<
+      const SeekableInputStream*,
+      std::shared_ptr<cache::CoalescedLoad>>>
+      fusedLoads_;
```
Comment (mbasmanova) [inline velox/dwio/dwrf/common/CachedBufferedInput.h:159]: fusedLoads_ -> coalescedLoads_

**Review by mbasmanova — COMMENTED** [review]
@oerling Orri, thank you for renaming FusedLoad to CoalescedLoad. It makes the code easier to read. What are we going to do about the gflags? Can we remove them altogether and pass these as constructor arguments? Do we need to keep some of them in which case can we put them all in a single place for easier tracking?

**Comment by oerling** [comment]
Edited the PR. Made an initialize method for cache entry.

File: velox/common/caching/AsyncDataCache.cpp [hunk 20]
```diff
@@ -231,47 +231,27 @@ bool FusedLoad::loadOrFuture(folly::SemiFuture<bool>* wait) {
...
-void FusedLoad::cancel() {
-  std::unique_ptr<folly::SharedPromise<bool>> promise;
-  {
-    std::lock_guard<std::mutex> l(mutex_);
-    if (state_ == LoadState::kLoading || state_ == LoadState::kCancelled) {
-      // Already cancelled or another thread is loading. If loading, we must let
-      // the load finish.
-      return;
+  } catch (std::exception& e) {
+    setEndState(LoadState::kCancelled);
```
Comment (mbasmanova) [inline velox/common/caching/AsyncDataCache.cpp:243]: This is a duplicate call that must have been added by mistake. Please, check.

File: velox/dwio/dwrf/common/CachedBufferedInput.cpp [hunk 21]
```diff
@@ -59,203 +77,321 @@ bool CachedBufferedInput::shouldPreload() {
...
+  // Extra requests made for preloadable regions that are larger then
+  // 'loadQuantum'.
+  std::vector<std::unique_ptr<CacheRequest>> extraRequests;
+  std::vector<CacheRequest*> storageLoad;
   for (auto& request : requests) {
+    cache::TrackingData trackingData;
+    if (!request.trackingId.empty()) {
+      trackingData = tracker_->trackingData(request.trackingId);
+    }
+    // Gather requently accessed items in a list and see if they should be
```
Comment (mbasmanova) [inline velox/dwio/dwrf/common/CachedBufferedInput.cpp:149]: typo: requently -> frequently

**Review by mbasmanova — APPROVED** [review]
@oerling Overall looks good, but I haven't reviewed the tests. It would be nice to move the new prefetch logic into a separate PR and add tests.
